### ISSUE TITLE
Error 403 Forbidden when creating `rancher2_cluster_v2` resource with a "Standard User"'s token

### ISSUE THREAD

**Issue by simu** [rancher/terraform-provider-rancher2#765]
## Context

I'm provisioning an RKE2 cluster in Rancher 2.6.0 with an API token belonging to an LDAP user with role "Standard User". I can create new RKE2 clusters in the Web interface with the same user without any issues. I can also use the same API token to create a `rancher2_cluster` resource with Terraform without any problems.

Terraform version: v1.0.7
Provider Rancher2 version: v1.20.1
Rancher version: 2.6.0, upgraded from 2.5.9 using Helm chart
Rancher Helm values:

```
auditLog:
  destination: hostPath
  hostPath: /var/log/auditlog/
  level: 2
  maxAge: 180
  maxBackup: 40
  maxSize: 250
hostname: test.rancher.<redacted>
ingress:
  tls:
    source: letsEncrypt
letsEncrypt:
  email: tech@<redacted>

```

Looking at the "Standard User" role definition from that Rancher instance I can see the following interesting snippets (from the YAML view):

```
rules:
[ ... snipped other rules ... ]
- apiGroups:
  - management.cattle.io
  resources:
  - clusters
  verbs:
  - create
[ .. snipped other rules .. ]
- apiGroups:
  - provisioning.cattle.io
  resources:
  - clusters
  verbs:
  - create
[ .. snipped other rules .. ]
```

It looks like standard users don't have permissions to `get` `clusters.provisioning.cattle.io` resources on our Rancher 2.6.0 instance. I can't determine whether that's the default "Standard User" role or if this has been customized in the past on our test instance.

## Actual behavior

Terraform fails with a permission error `Getting cluster V2 ID (fleet-default/rke2-terraform-test): Bad response statusCode [403]. Status [403 Forbidden].` Despite the Terraform error, the RKE2 cluster is created in Rancher and can be inspected in the Rancher web interface.

Find the full `terraform apply` output for the following Terraform below:

```
terraform {
  required_version = ">= 1.0"
  required_providers {
    rancher2 = {
      source  = "rancher/rancher2"
      version = "1.20.1"
    }
  }
}

resource "rancher2_cluster_v2" "cluster" {
  name = "rke2-terraform-test"

  kubernetes_version = "v1.21.4-rke2r2"
}
```

```
simon@phoenix:~/work/k8s/rke2-terraform-mve $ terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # rancher2_cluster_v2.cluster will be created
  + resource "rancher2_cluster_v2" "cluster" {
      + annotations                = (known after apply)
      + cluster_registration_token = (sensitive value)
      + cluster_v1_id              = (known after apply)
      + enable_network_policy      = (known after apply)
      + fleet_namespace            = "fleet-default"
      + id                         = (known after apply)
      + kube_config                = (sensitive value)
      + kubernetes_version         = "v1.21.4+rke2r2"
      + labels                     = (known after apply)
      + name                       = "rke2-terraform-test"
      + resource_version           = (known after apply)

      + rke_config {
          + additional_manifest   = (known after apply)
          + chart_values          = (known after apply)
          + machine_global_config = (known after apply)

          + etcd {
              + disable_snapshots      = (known after apply)
              + snapshot_retention     = (known after apply)
              + snapshot_schedule_cron = (known after apply)

              + s3_config {
                  + bucket                = (known after apply)
                  + cloud_credential_name = (known after apply)
                  + endpoint              = (known after apply)
                  + endpoint_ca           = (known after apply)
                  + folder                = (known after apply)
                  + region                = (known after apply)
                  + skip_ssl_verify       = (known after apply)
                }
            }

          + local_auth_endpoint {
              + ca_certs = (known after apply)
              + enabled  = (known after apply)
              + fqdn     = (known after apply)
            }

          + machine_pools {
              + annotations                  = (known after apply)
              + cloud_credential_secret_name = (known after apply)
              + control_plane_role           = (known after apply)
              + etcd_role                    = (known after apply)
              + labels                       = (known after apply)
              + name                         = (known after apply)
              + paused                       = (known after apply)
              + quantity                     = (known after apply)
              + worker_role                  = (known after apply)

              + machine_config {
                  + kind = (known after apply)
                  + name = (known after apply)
                }

              + rolling_update {
                  + max_surge       = (known after apply)
                  + max_unavailable = (known after apply)
                }

              + taints {
                  + effect = (known after apply)
                  + key    = (known after apply)
                  + value  = (known after apply)
                }
            }

          + machine_selector_config {
              + config = (known after apply)

              + machine_label_selector {
                  + match_labels = (known after apply)

                  + match_expressions {
                      + key      = (known after apply)
                      + operator = (known after apply)
                      + values   = (known after apply)
                    }
                }
            }

          + registries {
              + configs {
                  + auth_config_secret_name = (known after apply)
                  + ca_bundle               = (known after apply)
                  + hostname                = (known after apply)
                  + insecure                = (known after apply)
                  + tls_secret_name         = (known after apply)
                }

              + mirrors {
                  + endpoints = (known after apply)
                  + hostname  = (known after apply)
                  + rewrites  = (known after apply)
                }
            }

          + upgrade_strategy {
              + control_plane_concurrency = (known after apply)
              + worker_concurrency        = (known after apply)

              + control_plane_drain_options {
                  + delete_empty_dir_data                = (known after apply)
                  + disable_eviction                     = (known after apply)
                  + enabled                              = (known after apply)
                  + force                                = (known after apply)
                  + grace_period                         = (known after apply)
                  + ignore_daemon_sets                   = (known after apply)
                  + ignore_errors                        = (known after apply)
                  + skip_wait_for_delete_timeout_seconds = (known after apply)
                  + timeout                              = (known after apply)
                }

              + worker_drain_options {
                  + delete_empty_dir_data                = (known after apply)
                  + disable_eviction                     = (known after apply)
                  + enabled                              = (known after apply)
                  + force                                = (known after apply)
                  + grace_period                         = (known after apply)
                  + ignore_daemon_sets                   = (known after apply)
                  + ignore_errors                        = (known after apply)
                  + skip_wait_for_delete_timeout_seconds = (known after apply)
                  + timeout                              = (known after apply)
                }
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

rancher2_cluster_v2.cluster: Creating...
╷
│ Error: Getting cluster V2 ID (fleet-default/rke2-terraform-test): Bad response statusCode [403]. Status [403 Forbidden]. Body: [code=Forbidden, message=clusters.provisioning.cattle.io "rke2-terraform-test" is forbidden: User "u-n2qbinn2at" cannot get resource "clusters" in API group "provisioning.cattle.io" in the namespace "fleet-default"] from [https://test.rancher.<redacted>/v1/provisioning.cattle.io.clusters/fleet-default/rke2-terraform-test]
│ 
│   with rancher2_cluster_v2.cluster,
│   on rancher.tf line 1, in resource "rancher2_cluster_v2" "cluster":
│    1: resource "rancher2_cluster_v2" "cluster" {
│ 
╵
```

## Expected behavior

I can RKE2 clusters in Rancher with Terraform using the token of a "Standard User".

**Comment by simu** [comment]
Looking at the provider code I found the section in `waitForClusterV2State()` where we get the forbidden error when getting the V2 cluster ID: https://github.com/rancher/terraform-provider-rancher2/blob/6704c32e373dba57e30c1a4d6409d28ac23badb7/rancher2/resource_rancher2_cluster_v2.go#L216-L219

I performed a quick experiment locally, adding the `IsNotFound() || IsForbidden()` check after calling `getClusterV2ByID()`, arriving at the following implementation for `waitForClusterV2State()`:

```go
func waitForClusterV2State(c *Config, id, state string, interval time.Duration) (*ClusterV2, error) {
	if id == "" || state == "" {
		return nil, fmt.Errorf("Cluster V2 ID and/or condition is nil")
	}

	ctx, cancel := context.WithTimeout(context.Background(), interval)
	defer cancel()
	for {
		obj, err := getClusterV2ByID(c, id)
		if IsNotFound(err) || IsForbidden(err) {
			log.Printf("[INFO] Cluster ID %s not found.", id)
		} else if err != nil {
			return nil, fmt.Errorf("Getting cluster V2 ID (%s): %v", id, err)
		}
		if obj != nil {
			for i := range obj.Status.Conditions {
				if obj.Status.Conditions[i].Type == state {
					// Status of the condition, one of True, False, Unknown.
					if obj.Status.Conditions[i].Status == "Unknown" {
						break
					}
					if obj.Status.Conditions[i].Status == "True" {
						return obj, nil
					}
					// When cluster condition is false, retrying if it has been updated for last rancher2WaitFalseCond seconds
					lastUpdate, err := time.Parse(time.RFC3339, obj.Status.Conditions[i].LastUpdateTime)
					if err == nil && time.Since(lastUpdate) < rancher2WaitFalseCond*time.Second {
						break
					}
					return nil, fmt.Errorf("Cluster V2 ID %s: %s", id, obj.Status.Conditions[i].Message)
				}
			}
		}
		select {
		case <-time.After(rancher2RetriesWait * time.Second):
		case <-ctx.Done():
			return nil, fmt.Errorf("Timeout waiting for cluster V2 ID %s", id)
		}
	}
}
```

This change allows me to create RKE2 clusters using a "Standard User" token:

```
Terraform will perform the following actions:

  # rancher2_cluster_v2.cluster will be created
  + resource "rancher2_cluster_v2" "cluster" {
      + annotations                = (known after apply)
      + cluster_registration_token = (sensitive value)
      + cluster_v1_id              = (known after apply)
      + enable_network_policy      = (known after apply)
      + fleet_namespace            = "fleet-default"
      + id                         = (known after apply)
      + kube_config                = (sensitive value)
      + kubernetes_version         = "v1.21.4-rke2r2"
      + labels                     = (known after apply)
      + name                       = "rke2-terraform-test"
      + resource_version           = (known after apply)

      + rke_config {
          + additional_manifest   = (known after apply)
          + chart_values          = (known after apply)
          + machine_global_config = (known after apply)

          + etcd {
              + disable_snapshots      = (known after apply)
              + snapshot_retention     = (known after apply)
              + snapshot_schedule_cron = (known after apply)

              + s3_config {
                  + bucket                = (known after apply)
                  + cloud_credential_name = (known after apply)
                  + endpoint              = (known after apply)
                  + endpoint_ca           = (known after apply)
                  + folder                = (known after apply)
                  + region                = (known after apply)
                  + skip_ssl_verify       = (known after apply)
                }
            }

          + local_auth_endpoint {
              + ca_certs = (known after apply)
              + enabled  = (known after apply)
              + fqdn     = (known after apply)
            }

          + machine_pools {
              + annotations                  = (known after apply)
              + cloud_credential_secret_name = (known after apply)
              + control_plane_role           = (known after apply)
              + etcd_role                    = (known after apply)
              + labels                       = (known after apply)
              + name                         = (known after apply)
              + paused                       = (known after apply)
              + quantity                     = (known after apply)
              + worker_role                  = (known after apply)

              + machine_config {
                  + kind = (known after apply)
                  + name = (known after apply)
                }

              + rolling_update {
                  + max_surge       = (known after apply)
                  + max_unavailable = (known after apply)
                }

              + taints {
                  + effect = (known after apply)
                  + key    = (known after apply)
                  + value  = (known after apply)
                }
            }

          + machine_selector_config {
              + config = (known after apply)

              + machine_label_selector {
                  + match_labels = (known after apply)

                  + match_expressions {
                      + key      = (known after apply)
                      + operator = (known after apply)
                      + values   = (known after apply)
                    }
                }
            }

          + registries {
              + configs {
                  + auth_config_secret_name = (known after apply)
                  + ca_bundle               = (known after apply)
                  + hostname                = (known after apply)
                  + insecure                = (known after apply)
                  + tls_secret_name         = (known after apply)
                }

              + mirrors {
                  + endpoints = (known after apply)
                  + hostname  = (known after apply)
                  + rewrites  = (known after apply)
                }
            }

          + upgrade_strategy {
              + control_plane_concurrency = (known after apply)
              + worker_concurrency        = (known after apply)

              + control_plane_drain_options {
                  + delete_empty_dir_data                = (known after apply)
                  + disable_eviction                     = (known after apply)
                  + enabled                              = (known after apply)
                  + force                                = (known after apply)
                  + grace_period                         = (known after apply)
                  + ignore_daemon_sets                   = (known after apply)
                  + ignore_errors                        = (known after apply)
                  + skip_wait_for_delete_timeout_seconds = (known after apply)
                  + timeout                              = (known after apply)
                }

              + worker_drain_options {
                  + delete_empty_dir_data                = (known after apply)
                  + disable_eviction                     = (known after apply)
                  + enabled                              = (known after apply)
                  + force                                = (known after apply)
                  + grace_period                         = (known after apply)
                  + ignore_daemon_sets                   = (known after apply)
                  + ignore_errors                        = (known after apply)
                  + skip_wait_for_delete_timeout_seconds = (known after apply)
                  + timeout                              = (known after apply)
                }
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

2021-10-06T16:11:27.298+0200 [INFO]  provider: configuring client automatic mTLS
2021-10-06T16:11:27.328+0200 [DEBUG] provider: starting plugin: path=/home/simon/work/k8s/terraform-provider-rancher2/bin/terraform-provider-rancher2 args=[/home/simon/work/k8s/terraform-provider-rancher2/bin/terraform-provider-rancher2]
2021-10-06T16:11:27.328+0200 [DEBUG] provider: plugin started: path=/home/simon/work/k8s/terraform-provider-rancher2/bin/terraform-provider-rancher2 pid=354671
2021-10-06T16:11:27.328+0200 [DEBUG] provider: waiting for RPC address: path=/home/simon/work/k8s/terraform-provider-rancher2/bin/terraform-provider-rancher2
2021-10-06T16:11:27.352+0200 [INFO]  provider.terraform-provider-rancher2: configuring server automatic mTLS: timestamp=2021-10-06T16:11:27.352+0200
2021-10-06T16:11:27.381+0200 [DEBUG] provider.terraform-provider-rancher2: plugin address: address=/tmp/plugin131889839 network=unix timestamp=2021-10-06T16:11:27.381+0200
2021-10-06T16:11:27.381+0200 [DEBUG] provider: using plugin: version=5
rancher2_cluster_v2.cluster: Creating...
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 [DEBUG] setting computed for "rke_config" from ComputedKeys
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 [DEBUG] setting computed for "cluster_registration_token" from ComputedKeys
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 [DEBUG] setting computed for "labels" from ComputedKeys
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 [DEBUG] setting computed for "annotations" from ComputedKeys
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 [INFO] Creating Cluster V2 rke2-terraform-test
2021-10-06T16:11:27.463+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 Getting from  https://test.rancher.<redacted>/ping
2021-10-06T16:11:27.488+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:27 Time to get req:  25  ms
2021-10-06T16:11:29.131+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:29 [INFO] Cluster ID fleet-default/rke2-terraform-test forbidden.
2021-10-06T16:11:34.733+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:34 [INFO] Refreshing Cluster V2 fleet-default/rke2-terraform-test
2021-10-06T16:11:35.245+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:35 Getting from  https://test.rancher.<redacted>/ping
2021-10-06T16:11:35.266+0200 [DEBUG] provider.terraform-provider-rancher2: 2021/10/06 16:11:35 Time to get req:  20  ms
rancher2_cluster_v2.cluster: Creation complete after 9s [id=fleet-default/rke2-terraform-test]
2021-10-06T16:11:36.384+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-10-06T16:11:36.386+0200 [DEBUG] provider: plugin process exited: path=/home/simon/work/k8s/terraform-provider-rancher2/bin/terraform-provider-rancher2 pid=354671
2021-10-06T16:11:36.386+0200 [DEBUG] provider: plugin exited

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

**Comment by simu** [comment]
Testing my change (see previous comment) again, I now get the following error when trying to create an RKE2 cluster via Terraform using a token belonging to a "Standard User":

```
╷
│ Error: Getting cluster V2 ID (fleet-default/c-aged-breeze-2790): Getting cluster V2: Resource type [provisioning.cattle.io.cluster] can not be looked up by ID
│ 
│   with rancher2_cluster_v2.cluster,
│   on rancher.tf line 7, in resource "rancher2_cluster_v2" "cluster":
│    7: resource "rancher2_cluster_v2" "cluster" {
│ 
```

**Comment by rawmind0** [comment]
@simu , thanks for reporting this and looking into the code. 

The patch you made, was in the good direction but not enough. The provider clients should also be restarted to refresh RBAC until the standard user is able to read the cluster. I've included a fix at PR #770  https://github.com/rancher/terraform-provider-rancher2/pull/770/files#diff-f7df0599a821144eb6573b3b189435962d736692b3c14f72bd4e5244f94590a1R241

**Comment by simu** [comment]
Thanks for the feedback, I intentionally included my patch as a comment because I was not sure that it was sufficient to completely fix the issue.

**Comment by rawmind0** [comment]
PR https://github.com/rancher/terraform-provider-rancher2/pull/770 already merged. The fix will be included on next tf provider release, v1.21.0
